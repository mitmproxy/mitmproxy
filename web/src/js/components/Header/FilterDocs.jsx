import React, { Component } from 'react'
import { fetchApi } from "../../utils";


export default class FilterDocs extends Component {

    // @todo move to redux

    static xhr = null
    static doc = null

    constructor(props, context) {
        super(props, context)
        this.state = { doc: FilterDocs.doc }
    }

    componentDidMount() {
        if (!FilterDocs.xhr) {
            FilterDocs.xhr = fetchApi('/filter-help').then(response => response.json())
            FilterDocs.xhr.catch(() => {
                FilterDocs.xhr = null
            })
        }
        if (!this.state.doc) {
            FilterDocs.xhr.then(doc => {
                FilterDocs.doc = doc
                this.setState({ doc })
            })
        }
    }

    render() {
        const { doc } = this.state
        return !doc ? (
            <i className="fa fa-spinner fa-spin"></i>
        ) : (
            <table className="table table-condensed">
                <tbody>
                    {doc.commands.map(cmd => (
                        <tr key={cmd[1]} onClick={e => this.props.selectHandler(cmd[0].split(" ")[0] + " ")}>
                            <td>{cmd[0].replace(' ', '\u00a0')}</td>
                            <td>{cmd[1]}</td>
                        </tr>
                    ))}
                    <tr key="docs-link">
                        <td colSpan="2">
                            <a href="https://mitmproxy.org/docs/latest/concepts-filters/"
                                target="_blank">
                                <i className="fa fa-external-link"></i>
                            &nbsp; mitmproxy docs</a>
                        </td>
                    </tr>
                </tbody>
            </table>
        )
    }
}
