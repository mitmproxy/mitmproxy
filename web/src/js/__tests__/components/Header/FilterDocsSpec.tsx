import * as React from "react";
import FilterDocs from "../../../components/Header/FilterDocs";
import { enableFetchMocks } from "jest-fetch-mock";
import { render, screen, waitFor } from "../../test-utils";

enableFetchMocks();

test("FilterDocs Component", async () => {
    fetchMock.mockOnceIf(
        "./filter-help",
        JSON.stringify({
            commands: [
                ["cmd1", "foo"],
                ["cmd2", "bar"],
            ],
        }),
    );

    const { asFragment } = render(<FilterDocs selectHandler={() => 0} />);

    expect(asFragment()).toMatchSnapshot();
    await waitFor(() => screen.getByText("cmd1"));
    expect(asFragment()).toMatchSnapshot();
});
