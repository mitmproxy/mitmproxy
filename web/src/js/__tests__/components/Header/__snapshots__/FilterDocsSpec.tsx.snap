// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterDocs Component 1`] = `
<DocumentFragment>
  <i
    class="fa fa-spinner fa-spin"
  />
</DocumentFragment>
`;

exports[`FilterDocs Component 2`] = `
<DocumentFragment>
  <table
    class="table table-condensed"
  >
    <tbody>
      <tr>
        <td>
          cmd1
        </td>
        <td>
          foo
        </td>
      </tr>
      <tr>
        <td>
          cmd2
        </td>
        <td>
          bar
        </td>
      </tr>
      <tr>
        <td
          colspan="2"
        >
          <a
            href="https://mitmproxy.org/docs/latest/concepts-filters/"
            rel="noreferrer"
            target="_blank"
          >
            <i
              class="fa fa-external-link"
            />
            Â  mitmproxy docs
          </a>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
